### ISSUE TITLE
Division by zero in computeGradient of LossBinaryXENT

### ISSUE THREAD

**Issue by doerrhoefer-clemens** [deeplearning4j/deeplearning4j#3247]
#### Issue Description
In computeGradient function of LossBinaryXENT the denominator is calculated as.

**INDArray denominator = output.mul(output.rsub(1)); // output * (1-output)**

This term will be 0 if output == 1 or output == 0. Which leads to a division by zero and results in a NaN value.
 I am training a network where I have a sparse output of term distributions which leads to many output==0 when combining XENT Lossfunction with SOFTMAX activation function. Leading to a gradient of all NaN values. 

#### Version Information
I am Using deeplearning4j v0.8.0 with nd4j v0.8.0

**Comment by AlexDBlack** [comment]
BinaryXENT + Softmax generally aren't used together - typically (softmax+MCXENT) or (sigmoid+BinaryXENT) are used
However, this can still (in principle) occur with the latter (though, I have yet to see it in practice and I've used the latter pretty extensively). I'll leave this open though as we *might* want to consider an epsilon or conditional epsilon addition/subtraction (i.e., allow min of 1e-8 and max of 1-1e-8). I'm a bit hesitant to add that though due to the overhead and the fact that 99% of the time nets are fine without it...

Regularization (l1/l2 for weights and/or biases) can also help here.

**Comment by doerrhoefer-clemens** [comment]
I am aware that these are usually not used together. I am implementing Semantic Hashing described from this paper http://www.cs.utoronto.ca/~rsalakhu/papers/semantic_final.pdf. In this context it makes perfectly sense. I already extended RBM with Poisson sampling for pretraining and now the net needs to be finetuned.

**Comment by AlexDBlack** [comment]
I'm going to close this as a won't fix.
It's generally possible to handle underflow with tuning (regularization etc like I said).
It's possible to implement this as a custom loss function, however.
